src/internal/Observable.ts

```diff
@@ -392,4 +392,123 @@ export class Observable<T> implements Subscribable<T> {
     return pipeFromArray(operations)(this);
   }
+
+  /**
+   * Observable is async iterable, so it can be used in `for await` loop. This method
+   * of subscription is cancellable by breaking the for await loop. Although it's not
+   * recommended to use Observable's AsyncIterable contract outside of `for await`, if
+   * you're consuming the Observable as an AsyncIterable, and you're _not_ using `for await`,
+   * you can use the `throw` or `return` methods on the `AsyncGenerator` we return to
+   * cancel the subscription. Note that the subscription to the observable does not start
+   * until the first value is requested from the AsyncIterable.
+   *
+   * Functionally, this is equivalent to using a {@link concatMap} with an `async` function.
+   * That means that while the body of the `for await` loop is executing, any values that arrive
+   * from the observable source will be queued up, so they can be processed by the `for await`
+   * loop in order. So, like {@link concatMap} it's important to understand the speed your
+   * source emits at, and the speed of the body of your `for await` loop.
+   *
+   * ## Example
+   *
+   * ```ts
+   * import { interval } from 'rxjs';
+   *
+   * async function main() {
+   *  // Subscribe to the observable using for await.
+   *  for await (const value of interval(1000)) {
+   *    console.log(value);
+   *
+   *    if (value > 5) {
+   *      // Unsubscribe from the interval if we get a value greater than 5
+   *      break;
+   *    }
+   *  }
+   * }
+   *
+   * main();
+   * ```
+   */
+  [Symbol.asyncIterator](): AsyncGenerator<T, void, void> {
+    let subscription: Subscription | undefined;
+    let hasError = false;
+    let error: any;
+    let completed = false;
+    const values: T[] = [];
+    const deferreds: [(value: any) => void, (reason: any) => void][] = [];
+
+    const handleError = (err: any) => {
+      hasError = true;
+      error = err;
+      while (deferreds.length) {
+        deferreds.shift()![1](err);
+      }
+    };
+
+    const handleComplete = () => {
+      completed = true;
+      while (deferreds.length) {
+        deferreds.shift()![0]({ value: undefined, done: true });
+      }
+    };
+
+    return {
+      next: (): Promise<IteratorResult<T>> => {
+        if (!subscription) {
+          // We only want to start the subscription when the user starts iterating.
+          subscription = this.subscribe({
+            next: (value) => {
+              if (deferreds.length) {
+                deferreds.shift()![0]({ value, done: false });
+              } else {
+                values.push(value);
+              }
+            },
+            error: handleError,
+            complete: handleComplete,
+          });
+        }
+
+        // If we already have some values in our buffer, we'll return the next one.
+        if (values.length) {
+          return Promise.resolve({ value: values.shift()!, done: false });
```

Comment:
This is arguably a violation of the AsyncIterator protocol in the case that the value produced by the observable is a Promise - it will result in you seeing a Promise as the value in a for-await loop, as in

```js
for await (let item of observable) {
  assert(item instanceof Promise); // not possible with built-in async iterators
}
```

The language generally tries to prevent this from happening - that is, if you `yield` a Promise in an AsyncGenerator it will be `await`'d, and if you `for await` over a sync iterable containing Promises then the Promises will be `await`'d before each iteration of the loop.

Also the [async iterator helpers proposal](https://github.com/tc39/proposal-async-iterator-helpers) is intending to add helpers like `map` (etc) to the async iterator prototype, and the values returned by the callbacks passed to the helpers will be `await`d, which similarly prevents you from observing a Promise in the `value` slot of an iteration result from an AsyncIterator.

---

Now, this is kind of a weird case. Nothing in the language formally prevents you from putting a Promise there, it just avoids doing so itself. There was a [whole debate](https://docs.google.com/presentation/d/1U6PivKbFO0YgoFlrYB82MtXf1ofCp1xSVOODOvranBM/edit#slide=id.p) about how to handle this during the design of async iterables. But the upshot was, to quote one of the slides about the design (option 3) which was ultimately chosen,

> you could manually assemble an async iterator whose next() returns promises whose value is a promise, and we wouldn't unwrap them, and consumers would see the inner promise. That's fine; it's a contract violation.

That's what's happening here.

---

If you want to avoid the contract violation and match the rest of the language better, you can do something like `return Promise.resolve(values.shift()).then(value => ({ value, done: false }))` (though I haven't thought about the error case - probably you need to unsubscribe in that case).

Reply:
I think you're conflating Async Generators and AsyncIterable here. While it's true that yielding a Promise in a generator will cause the promise to resolve before the value comes out in the async iteration result, there's no hard-and-fast rule that says "you can't have an AsyncIterator<Promise<T>>". If it wasn't allowed to work that way, then the `for await` loop would resolve the value.

In any case, this is a pretty extreme corner case, as I've never seen anyone create an `Observable<Promise<T>>`, and it seems even more unlikely they'd do that with the intent to interop to an async function, unless they REALLY wanted to do that for some reason.

I don't really want to add code that will further deoptimize this unless there's a compelling reason.

As far as async iterator helpers, I think I answered that in another comment.

Reply:
> I think you're conflating AsyncGenerator semantics and AsyncIterable consumption here. 

Well, yes and no. There's another case, which is, sync iterators get wrapped up into async iterators by `for await` loops, and the wrapper `await`s any promises yielded by the sync iterator specifically so as to not expose Promises to `for await`.

And right now async generators and the async-from-sync wrapper are the only things in the language which produce async iterators, so everything (all two things) in the language ensures this property is maintained. The point about async iterator helpers was just to say that the helpers will add a new built-in way of producing async iterators (from other iterators), and they'll maintain this property as well (though mostly by necessity - you really want to be able to pass an async function to `.map` on an async iterator, which means you have to await its result anyway).

But yes, as you say nothing in the language stops you from doing this when constructing async iterators manually. Perfectly reasonable to decide it's not worth the overhead in order to maintain consistency in this edge case.